theano/tensor/signal/downsample.py

```diff
@@ -268,78 +268,123 @@ def grad(self, inp, grads):
                                         ignore_border=self.ignore_border,
                                         st=self.st, padding=self.padding)(
                                             x, maxout, gz)]
-
+    def c_headers(self):
+        #import ipdb; ipdb.set_trace()
+        #libs = super(DownsampleFactorMax, self).c_headers()
+        return ['<algorithm>']
+    
     def c_code(self, node, name, inp, out, sub):
         # No implementation is currently for the case where
         # the stride size and the pooling size are different.
         # An exception is raised for such a case.
-        if self.ds != self.st or self.padding != (0, 0):
-            raise theano.gof.utils.MethodNotDefined()
+        #if self.ds != self.st or self.padding != (0, 0):
+        #    raise theano.gof.utils.MethodNotDefined()
         x, = inp
         z, = out
         fail = sub['fail']
         ignore_border = int(self.ignore_border)
         ds0, ds1 = self.ds
+        st0, st1 = self.st
+        pd0, pd1 = self.padding 
         return """
         int typenum = PyArray_ObjectType((PyObject*)%(x)s, 0);
-        int x_shp0_usable;
-        int x_shp1_usable;
-        int z_shp0, z_shp1;
+        int z_r, z_c; // shape of the output
+        int r, c; // shape of the input
+        
         if(PyArray_NDIM(%(x)s)!=4)
         {
             PyErr_SetString(PyExc_ValueError, "x must be a 4d ndarray");
             %(fail)s;
         }
-        z_shp0 = PyArray_DIMS(%(x)s)[2] / %(ds0)s;
-        z_shp1 = PyArray_DIMS(%(x)s)[3] / %(ds1)s;
+        r = PyArray_DIMS(%(x)s)[2];
+        c = PyArray_DIMS(%(x)s)[3];
+        r += %(pd0)s * 2;
+        c += %(pd1)s * 2;
+        
         if (%(ignore_border)s)
         {
-            x_shp0_usable = z_shp0 * %(ds0)s;
-            x_shp1_usable = z_shp1 * %(ds1)s;
+            z_r = (r - %(ds0)s) / %(st0)s + 1;
+            z_c = (c - %(ds1)s) / %(st1)s + 1;
```

Comment:
You should raise an exception if `ignore_border` is true but `padding` is not `(0, 0)`.